### PR TITLE
Sundials 4.0.2

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

**Comment by FFroehlich** [comment]
okay, this is going to be a large PR

File: src/sundials_matrix_wrapper.cpp [hunk 1]
```diff
@@ -1,126 +1,83 @@
...
-
-
-
-DlsMatWrapper::DlsMatWrapper(long int M, long int N)
-    : matrix(NewDenseMat(M, N))
+    
+SUNMatrixWrapper::SUNMatrixWrapper(int M, int N)
+: matrix(SUNDenseMatrix(M, N))
 {
-    if((M*N > 0) && !matrix)
```
Comment (dweindl) [inline src/sundials_matrix_wrapper.cpp]: That gave me a compiler warning. Multiplication in boolean context or something like that.

**Comment by FFroehlich** [comment]
I think most of the heavy lifting is done, but some issues still remain:

- [ ] create RAII wrappers for sundials classes `SUNNonlinearSolver`  and `SUNLinearSolver`
- [ ] implement creation `SUNNonlinearSolver` (based on `iter`) and `SUNLinearSolver` (based on `linsol`) instances in `solver_cvodes` and `solver_idas` interfaces and remove deprecated function
- [ ] replace functionality `solver::ReInit` and `solver::sensReInit`, `solver::reInitB` and `solver::quadReInitB` by resetting order to zero and updating respective vectors in the solver memory (see #398) 

@dweindl could you take care of the first two?

**Comment by dweindl** [comment]
@FFroehlich Compiles now and tests pass, except for model_robertson_test. Will clean up more.

**Comment by dweindl** [comment]
While we are on it: sundials 4.0.2 was released last week ;)

File: src/solver.cpp [hunk 2]
```diff
@@ -213,154 +212,173 @@ void Solver::getDiagnosisB(const int it, ReturnData *rdata, int which) const {
...
+        throw AmiException("Invalid choice of solver: %d", static_cast<int>(linsol));
+    }
+}
+
+void Solver::initializeNonLinearSolverB(AmiVector *xB, const int which, Model *model)
+{
+    switch(iter) {
+    case NonlinearSolverIteration::newton:
+        switch(ism) {
+        case InternalSensitivityMethod::staggered:
```
Comment (FFroehlich) [inline src/solver.cpp]: We never use backwards integration + sensitivities, so this should not be necessary.

**Comment by FFroehlich** [comment]
Okay, I think everything but the events stuff should now be working

**Comment by FFroehlich** [comment]
okay `example_adjoint` is working now. Still need to implement the event stuff for IDAS and after that we probably will have to do some general cleanup, but it looks like might be able to merge this PR soon.

**Comment by dweindl** [comment]
Great! Will start looking into the cleaning part.

**Comment by dweindl** [comment]
Tests are passing for you? I have 5 failing here.

**Pull request merged**